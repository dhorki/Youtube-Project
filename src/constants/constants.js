export const styles = {
  colors: {
    light: {
      appIcon: '#ff0200',
      background: '#F9F9F9',
      barBackground: '#ffffff',
      primary: '#030303',
      secondary: '#606060',
      linkActive: '#ff0200',
      linkHoverBackground: '#ececec',
      inputBackground: '#ffffff',
      inputBorder: '#cccccc',
      searchButtonBackground: '#f8f8f8',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
    dark: {
      appIcon: '#ff0200',
      background: '#181818',
      barBackground: '#212121',
      primary: '#ffffff',
      secondary: '#aaaaaa',
      linkActive: '#ff0200',
      linkHoverBackground: '#303030',
      inputBackground: '#121212',
      inputBorder: '#303030',
      searchButtonBackground: '#323232',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
  },
  props: {
    transitionTime: '0.4s',
    sidebarWidth: '150px',
  },
};

export const screenSizes = {
  xsmall: '675px',
  small: '830px',
  mid: '1200px',
};

export const mainViewTypes = {
  searchGallery: 'searchGallery',
  favoritesGallery: 'favoritesGallery',
  videoDetails: 'videoDetails',
};

export const environmentActions = {
  setTheme: 'setTheme',
  setSidebarShow: 'setSidebarShow',
  setSearchQuery: 'setSearchQuery',
  setModalShow: 'setModalShow',
  setUser: 'setUser',
};

export const modalTypes = {
  login: 'login',
  register: 'register',
  resetPassword: 'reset-password',
  none: 'none',
};

export const resetSearchQueryAction = {
  type: environmentActions.setSearchQuery,
  payload: {
    q: '',
  },
};

export const hideModalAction = {
  type: environmentActions.setModalShow,
  payload: {
    modalShow: modalTypes.none,
  },
};
