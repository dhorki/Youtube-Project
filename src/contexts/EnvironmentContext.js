import { createContext } from 'react';

export const EnvironmentContext = createContext();
