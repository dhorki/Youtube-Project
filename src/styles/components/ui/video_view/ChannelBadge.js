import styled from 'styled-components';

export const StyledChannelBadge = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  margin: 0px;
  padding: 0px;
  height: 100%;
  color: ${({ theme }) => theme.primary};

  > .badge {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;

    > img {
      width: 48px;
      height: 48px;
      border-radius: 50%;
      overflow: hidden;
      margin: 0 10px;
    }

    > div {
      display: flex;
      flex-direction: column;
      justify-content: left;
      align-items: left;

      > .title {
        margin: 0px;
        font-size: 15px;
        color: ${({ theme }) => theme.primary};
      }

      > .subscribers {
        margin: 5px 0;
        font-size: 13px;
        color: ${({ theme }) => theme.secondary};
      }
    }
  }
`;
