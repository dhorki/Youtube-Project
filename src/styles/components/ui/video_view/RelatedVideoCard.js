import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { styles } from '../../../../constants/constants';

export const StyledRelatedVideoCard = styled(Link)`
  display: flex;
  flex-direction: row;
  width: 360px;

  margin-left: 5px;
  margin-bottom: 10px;
  background-color: ${({ theme }) => theme.background};
  color: ${({ theme }) => theme.primary};
  text-decoration: none;

  transition: background-color ${styles.props.transitionTime} ease;

  > img {
    width: 168px;
    height: 94px;
  }

  > div {
    display: flex;
    flex-direction: column;
    width: 226px;
    margin-left: 10px;

    > .videocard-title {
      color: ${({ theme }) => theme.primary};
      font-size: 14px;
      font-weight: bold;
    }

    > .videocard-channel-title {
      margin-top: 10px;
      color: ${({ theme }) => theme.secondary};
      font-size: 14px;
    }
  }
`;
