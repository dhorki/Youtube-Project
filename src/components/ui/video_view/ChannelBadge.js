import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { StyledChannelBadge } from '../../../styles/components/ui/video_view/ChannelBadge';
import { useFetch } from '../../../hooks/useFetch';
import { intToMagnitude } from '../../../helpers/formatters';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const ChannelBadge = ({ id }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const url =
    'https://www.googleapis.com/youtube/v3/channels?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet,statistics&' +
    `id=${id}&` +
    'maxResults=24';

  const { data, loading, error } = useFetch(url);

  let image = '',
    title = '#?',
    subscriberCount = 0;
  if (!loading && !error) {
    const item = data?.items[0];
    image = item.snippet.thumbnails.default.url;
    title = item.snippet.title;
    subscriberCount = item.statistics.subscriberCount;
  }

  return (
    <StyledChannelBadge theme={theme}>
      {loading ? null : (
        <div className="badge">
          {!error ? <img src={image} alt={title} /> : null}
          <div>
            <p className="title">{title}</p>
            <p className="subscribers">{intToMagnitude(subscriberCount)} subscribers</p>
          </div>
        </div>
      )}
    </StyledChannelBadge>
  );
};

ChannelBadge.propTypes = {
  id: PropTypes.string.isRequired,
};
