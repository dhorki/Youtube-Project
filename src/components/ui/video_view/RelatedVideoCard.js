import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { timeSince } from '../../../helpers/formatters';
import { decodeHtml } from '../../../helpers/decodeHtml';
import { StyledRelatedVideoCard } from '../../../styles/components/ui/video_view/RelatedVideoCard';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const RelatedVideoCard = ({ video }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const { id, snippet } = video;
  const { videoId } = id;
  const { title, channelTitle, thumbnails, publishTime } = snippet;
  const { url } = thumbnails.medium;
  const d = new Date(publishTime);

  const upperCaseCharsCount = title.length - title.replaceAll(/[A-Z]/g, '').length;
  const limitLength = upperCaseCharsCount > 50 ? 50 : upperCaseCharsCount > 30 ? 60 : 70;

  const fixedTitle =
    title.length >= limitLength ? title.slice(0, limitLength) + '...' : title;

  return (
    <StyledRelatedVideoCard
      className="animate__animated animate__fadeIn"
      theme={theme}
      to={`/${videoId}`}
    >
      <img src={url} alt="video" />
      <div>
        <div className="videocard-title">{decodeHtml(fixedTitle)}</div>
        <div className="videocard-channel-title">
          {decodeHtml(channelTitle)}
          <br />
          {timeSince(d)} ago.
        </div>
      </div>
    </StyledRelatedVideoCard>
  );
};

RelatedVideoCard.propTypes = {
  video: PropTypes.object.isRequired,
};
