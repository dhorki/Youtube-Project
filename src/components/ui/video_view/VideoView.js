import React, { useState } from 'react';
import { RiLiveLine } from 'react-icons/ri';
import { FaThumbsUp } from 'react-icons/fa';
import { IoTimerOutline } from 'react-icons/io5';
import {
  formatYoutubeDate,
  intToMagnitude,
  numberWithCommas,
  youtubeDurationToTime,
} from '../../../helpers/formatters';
import PropTypes from 'prop-types';
import { useFetchYoutubeVideos } from '../../../hooks/useFetchYoutubeVideos';
import { useTheme } from '../../../hooks/useTheme';

import { StyledVideoView } from '../../../styles/components/ui/video_view/VideoView';
import { LoadingAnimation } from '../loading/LoadingAnimation';
import { ChannelBadge } from './ChannelBadge';
import { ErrorText } from '../error/ErrorText';

export const VideoView = ({ id }) => {
  const { getTheme } = useTheme();
  const theme = getTheme();

  const { data, loading, error } = useFetchYoutubeVideos(
    process.env.REACT_APP_YOUTUBE_API_KEY,
    [id]
  );

  const [descriptionCollapsed, setDescriptionCollapsed] = useState(true);

  const iframeAllow =
    'accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture';

  // const { items } = require('../../../mocks/youtube-video-mock.json');
  let snippet, contentDetails, statistics;
  if (!loading && !error) {
    console.log(data);
    const item = data?.items[0];
    snippet = item.snippet;
    contentDetails = item.contentDetails;
    statistics = item.statistics;
  }

  return (
    <StyledVideoView className="animate__animated animate__fadeIn" theme={theme}>
      {loading ? (
        <LoadingAnimation />
      ) : error ? (
        <ErrorText error={error} />
      ) : (
        <>
          <div className="embedded-video">
            <iframe
              title={id}
              src={`//www.youtube.com/embed/${id}`}
              frameBorder="0"
              allow={iframeAllow}
              allowFullScreen="allowfullscreen"
            />
          </div>
          <p className="view-video-title">{snippet.title}</p>
          <div className="view-video-stats">
            <p className="view-video-stats-left">
              <span>{numberWithCommas(statistics.viewCount)}</span> views -{' '}
              <span>{formatYoutubeDate(snippet.publishedAt)}</span>
            </p>
            <p className="view-video-stats-right">
              <span>
                <FaThumbsUp /> {intToMagnitude(statistics.likeCount)}
              </span>{' '}
              <span>
                <IoTimerOutline />{' '}
                {snippet.liveBroadcastContent === 'live' ? (
                  <RiLiveLine className="live-icon" />
                ) : (
                  youtubeDurationToTime(contentDetails.duration)
                )}
              </span>
            </p>
          </div>
          <hr />
          <div className="view-video-channel-favorite">
            <ChannelBadge id={snippet.channelId} />
            <button>ADD TO FAVORITES</button>
          </div>
          <div className="view-video-description-box">
            <p
              className={`view-video-description ${
                descriptionCollapsed ? 'collapsed' : ''
              }`}
              dangerouslySetInnerHTML={{
                __html: snippet.description.replaceAll('\n', '<br />'),
              }}
            />
            <button
              className="view-video-description-toggle"
              onClick={() => setDescriptionCollapsed(!descriptionCollapsed)}
            >
              SHOW {descriptionCollapsed ? 'MORE' : 'LESS'}
            </button>
          </div>
        </>
      )}
    </StyledVideoView>
  );
};

VideoView.propTypes = {
  id: PropTypes.string.isRequired,
};
