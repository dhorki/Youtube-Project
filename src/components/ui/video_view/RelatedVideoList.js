import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { StyledRelatedVideoList } from '../../../styles/components/ui/video_view/RelatedVideoList';
import { useFetch } from '../../../hooks/useFetch';
import { RelatedVideoCard } from './RelatedVideoCard';
import { LoadingAnimation } from '../loading/LoadingAnimation';
import { ErrorText } from '../error/ErrorText';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const RelatedVideoList = ({ relatedTo }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const url =
    'https://www.googleapis.com/youtube/v3/search?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet&' +
    'type=video&' +
    'order=relevance&' +
    `relatedToVideoId=${relatedTo}&` +
    'maxResults=24';

  const { data, loading, error } = useFetch(url);
  // const { data, loading, error } = { data: [], loading: true, error: null };

  let filteredItems = [];

  if (!loading && !error) {
    filteredItems = data.items?.filter(
      (item) => item.id.kind.includes('youtube#video') && item.snippet
    );
  }

  return (
    <StyledRelatedVideoList theme={theme}>
      <h3>Related Videos</h3>
      {loading ? (
        <LoadingAnimation />
      ) : error ? (
        <ErrorText error={error} />
      ) : !filteredItems || filteredItems.length === 0 ? (
        'No Related videos.'
      ) : (
        filteredItems.map((video) => (
          <RelatedVideoCard key={video.id.videoId} video={video} />
        ))
      )}
    </StyledRelatedVideoList>
  );
};

RelatedVideoList.propTypes = {
  relatedTo: PropTypes.string.isRequired,
};
