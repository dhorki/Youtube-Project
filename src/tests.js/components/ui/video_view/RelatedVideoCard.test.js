import React from 'react';
import { Router } from 'react-router-dom';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { createBrowserHistory } from 'history';
import { styles } from '../../../../constants/constants';
import { EnvironmentContext } from '../../../../contexts/EnvironmentContext';
import { RelatedVideoCard } from '../../../../components/ui/video_view/RelatedVideoCard';

describe('RelatedVideoCard', () => {
  const environment = {
    theme: styles.colors.light,
    sidebarShow: false,
    q: '',
  };

  it('should render correctly', () => {
    const date = new Date();
    date.setDate(date.getDate() - 10);
    const publishTime = date.toISOString().replace(/\.\d+/g, '');

    const video = {
      id: 'videoId',
      snippet: {
        title: 'Video title',
        channelTitle: 'Channel title',
        description: 'Video description',
        thumbnails: {
          medium: {
            url: 'https://www.youtube.com/watch?v=videoId',
          },
        },
        publishTime: publishTime,
      },
    };

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <Router history={createBrowserHistory()}>
          <RelatedVideoCard video={video} />
        </Router>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly at different title lengths', () => {
    const date = new Date();
    date.setDate(date.getDate() - 10);
    const publishTime = date.toISOString().replace(/\.\d+/g, '');

    const titleHighUpperCaseChars = ''.padEnd(51, 'A') + ''.padEnd(100, 'a');
    const titleMidUpperCaseChars = ''.padEnd(31, 'B') + ''.padEnd(100, 'b');
    const titleLowUpperCaseChars = ''.padEnd(30, 'C') + ''.padEnd(100, 'c');

    const video1 = {
      id: 'videoId',
      snippet: {
        title: titleHighUpperCaseChars,
        channelTitle: 'Channel title',
        description: 'Video description',
        thumbnails: {
          medium: {
            url: 'https://www.youtube.com/watch?v=videoId',
          },
        },
        publishTime: publishTime,
      },
    };

    const wrapper1 = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <Router history={createBrowserHistory()}>
          <RelatedVideoCard video={video1} />
        </Router>
      </EnvironmentContext.Provider>
    );

    expect(wrapper1.find('.videocard-title').text()).toBe(
      titleHighUpperCaseChars.slice(0, 50) + '...'
    );

    const video2 = {
      id: 'videoId',
      snippet: {
        title: titleMidUpperCaseChars,
        channelTitle: 'Channel title',
        description: 'Video description',
        thumbnails: {
          medium: {
            url: 'https://www.youtube.com/watch?v=videoId',
          },
        },
        publishTime: publishTime,
      },
    };

    const wrapper2 = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <Router history={createBrowserHistory()}>
          <RelatedVideoCard video={video2} />
        </Router>
      </EnvironmentContext.Provider>
    );

    expect(wrapper2.find('.videocard-title').text()).toBe(
      titleMidUpperCaseChars.slice(0, 60) + '...'
    );

    const video3 = {
      id: 'videoId',
      snippet: {
        title: titleLowUpperCaseChars,
        channelTitle: 'Channel title',
        description: 'Video description',
        thumbnails: {
          medium: {
            url: 'https://www.youtube.com/watch?v=videoId',
          },
        },
        publishTime: publishTime,
      },
    };

    const wrapper3 = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <Router history={createBrowserHistory()}>
          <RelatedVideoCard video={video3} />
        </Router>
      </EnvironmentContext.Provider>
    );

    expect(wrapper3.find('.videocard-title').text()).toBe(
      titleLowUpperCaseChars.slice(0, 70) + '...'
    );

    // wrapper.setProps({
    //   video: {
    //     ...video,
    //     snippet: {
    //       ...video.snippet,
    //       title: titleMidUpperCaseChars,
    //     },
    //   },
    // });
    // wrapper.update();

    // expect(wrapper.find('.videocard-title').text()).toBe(''.padEnd(30, 'B') + '...');

    // wrapper.setProps({
    //   video: {
    //     ...video,
    //     snippet: {
    //       ...video.snippet,
    //       title: titleLowUpperCaseChars,
    //     },
    //   },
    // });
    // wrapper.update();

    // expect(wrapper.find('.videocard-title').text()).toBe(titleLowUpperCaseChars);
  });
});
