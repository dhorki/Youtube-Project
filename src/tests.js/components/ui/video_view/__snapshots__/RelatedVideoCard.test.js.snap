// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideoCard should render correctly 1`] = `
<a
  className="sc-bczRLJ jnCrbq animate__animated animate__fadeIn"
  href="/undefined"
  onClick={[Function]}
  theme={
    Object {
      "appIcon": "#ff0200",
      "background": "#F9F9F9",
      "barBackground": "#ffffff",
      "inputBackground": "#ffffff",
      "inputBorder": "#cccccc",
      "linkActive": "#ff0200",
      "linkHoverBackground": "#ececec",
      "primary": "#030303",
      "searchButtonBackground": "#f8f8f8",
      "secondary": "#606060",
    }
  }
>
  <img
    alt="video"
    src="https://www.youtube.com/watch?v=videoId"
  />
  <div>
    <div
      className="videocard-title"
    >
      Video title
    </div>
    <div
      className="videocard-channel-title"
    >
      Channel title
      <br />
      10 days 10 hours
       ago.
    </div>
  </div>
</a>
`;
